trieste/space.py

```diff
@@ -331,3 +332,138 @@ def __mul__(self, other: Box) -> Box:
+        :param upper: The upper (inclusive) bounds of the Ordinal Space. Must have shape [D]
+            for positive D, and if a tensor, must have float type.
+        :param stepsizes: The step size of each dimension of the Ordinal Space. Must have
+            shape [D] for positive D, and if a tensor, must have float type.
+        :raise ValueError (or InvalidArgumentError): If any of the following are true:
+
+            - ``lower``, ``upper`` and ``stepsizes`` have invalid shapes.
+            - ``lower``, ``upper`` and ``stepsizes`` do not have the same floating point type.
+            - ``upper`` is not greater than ``lower`` across all dimensions.
+        """
+
+        super().__init__(lower, upper)
+        tf.debugging.assert_shapes([(lower, ["D"]), (stepsizes, ["D"])])
+        tf.assert_rank(stepsizes, 1)
+
+        tf.debugging.assert_positive(len(stepsizes), message="stepsizes cannot be empty")
+
+        if isinstance(stepsizes, Sequence):
+            self._stepsizes = tf.constant(stepsizes, dtype=tf.float64)
+        else:
+            self._stepsizes = tf.convert_to_tensor(stepsizes)
+
+            tf.debugging.assert_same_float_dtype([self._lower, self._stepsizes])
+
+    def __repr__(self) -> str:
+        """"""
+        return f"OrdinalSearchSpace({self._lower!r}, {self._upper!r}, {self._stepsizes!r})"
+
+    def sample(self, num_samples: int) -> TensorType:
+        """
+        :param num_samples: The number of points to sample from this search space.
+        :return: ``num_samples`` i.i.d. random points, sampled uniformly, and without replacement,
+            from this search space.
+        """
+        samples = super().sample(num_samples)
+        samples = tf.round(samples / self._stepsizes) * self._stepsizes + 0
+
+        return samples
+
+    @property
+    def stepsizes(self) -> TensorType:
+        """the step sizes of the ordinal space"""
+        return self._stepsizes
+
+    def __contains__(self, value: TensorType) -> bool | TensorType:
+        """
+        Return `True` if ``value`` is a member of this search space, else `False`. A point is a
+        member if all of its coordinates lie in the closed intervals bounded by the lower and upper
+        bounds and spaced by some step sizes.
+
+        :param value: A point to check for membership of this :class:`SearchSpace`.
+        :return: `True` if ``value`` is a member of this search space, else `False`. May return a
+            scalar boolean `TensorType` instead of the `bool` itself.
+        :raise ValueError (or InvalidArgumentError): If ``value`` has a different dimensionality
+            from the search space.
+        """
+        if not shapes_equal(value, self._lower):
+            raise ValueError(
+                f"value must have same dimensionality as search space: {self._lower.shape},"
+                f" got shape {value.shape}"
+            )
+
+        isclose_array = np.vectorize(np.isclose)
+
```

Comment:
Can this not be done in tf?

Reply:
they have `tf.experimental.numpy.isclose ` in tensorflow. and this function vectorization might be possible with `tf.map_fn`. what do you think?

Reply:
yep give it a go. We dont want numpy

Reply:
👌🏼